lucene/core/src/java/org/apache/lucene/index/IndexWriter.java

```diff
@@ -1554,20 +1554,24 @@ public long deleteDocuments(Query... queries) throws IOException {
       }
     }
 
-    try {
-      long seqNo = docWriter.deleteQueries(queries);
-      if (seqNo < 0) {
-        seqNo = -seqNo;
-        processEvents(true);
-      }
-
-      return seqNo;
+    try (Closeable finalizer = acquireModificationLease()) {
+      return maybeProcessEvents(docWriter.deleteQueries(queries));
     } catch (VirtualMachineError tragedy) {
       tragicEvent(tragedy, "deleteDocuments(Query..)");
       throw tragedy;
     }
   }
 
+  private Closeable acquireModificationLease() {
+    assert Thread.holdsLock(this) == false : "Can't acquire modification lock while holding the IW lock";
```

Comment:
Can/should we also assert the reverse?  I.e. whenever we `sync` on `this`, `assert` that the `modificationLease` is not held?  They are mutually exclusive?